File: Cyclically Rotating a Grid/Cyclically Rotating a Grid.cpp
// Solution for Cyclically Rotating a Grid in CPP

class Solution {
public:
    vector<vector<int>> rotateGrid(vector<vector<int>>& grid, int k) {

        int rows = grid.size();
        int cols = grid[0].size();

        // Number of concentric layers in matrix
        int totalLayers = min(rows, cols) / 2;

        // Process each layer independently
        for (int layer = 0; layer < totalLayers; layer++) {

            int topRow = layer;
            int leftCol = layer;

            int bottomRow = rows - layer - 1;
            int rightCol = cols - layer - 1;

            vector<int> layerElements;

            // -----------------------------------
            // Extract current layer elements
            // in traversal order
            // -----------------------------------

            // Top row (left -> right)
            for (int col = leftCol; col < rightCol; col++) {
                layerElements.push_back(grid[topRow][col]);
            }

            // Right column (top -> bottom)
            for (int row = topRow; row < bottomRow; row++) {
                layerElements.push_back(grid[row][rightCol]);
            }

            // Bottom row (right -> left)
            for (int col = rightCol; col > leftCol; col--) {
                layerElements.push_back(grid[bottomRow][col]);
            }

            // Left column (bottom -> top)
            for (int row = bottomRow; row > topRow; row--) {
                layerElements.push_back(grid[row][leftCol]);
            }

            int layerSize = layerElements.size();

            // Effective rotation needed
            int rotationSteps = k % layerSize;

            // Counter-clockwise cyclic rotation
            rotate(layerElements.begin(),
                   layerElements.begin() + rotationSteps,
                   layerElements.end());

            int currentIndex = 0;

            // -----------------------------------
            // Fill rotated values back into grid
            // -----------------------------------

            // Top row
            for (int col = leftCol; col < rightCol; col++) {
                grid[topRow][col] = layerElements[currentIndex++];
            }

            // Right column
            for (int row = topRow; row < bottomRow; row++) {
                grid[row][rightCol] = layerElements[currentIndex++];
            }

            // Bottom row
            for (int col = rightCol; col > leftCol; col--) {
                grid[bottomRow][col] = layerElements[currentIndex++];
            }

            // Left column
            for (int row = bottomRow; row > topRow; row--) {
                grid[row][leftCol] = layerElements[currentIndex++];
            }
        }

        return grid;
    }
};

File: Cyclically Rotating a Grid/Cyclically Rotating a Grid.py
# Solution for Cyclically Rotating a Grid in PY

class Solution:
    def rotateGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:

        rows = len(grid)
        cols = len(grid[0])

        # Number of layers
        total_layers = min(rows, cols) // 2

        for layer in range(total_layers):

            top_row = layer
            left_col = layer

            bottom_row = rows - layer - 1
            right_col = cols - layer - 1

            layer_elements = []

            # Top row
            for col in range(left_col, right_col):
                layer_elements.append(grid[top_row][col])

            # Right column
            for row in range(top_row, bottom_row):
                layer_elements.append(grid[row][right_col])

            # Bottom row
            for col in range(right_col, left_col, -1):
                layer_elements.append(grid[bottom_row][col])

            # Left column
            for row in range(bottom_row, top_row, -1):
                layer_elements.append(grid[row][left_col])

            layer_size = len(layer_elements)

            # Effective rotation
            rotation_steps = k % layer_size

            # Counter-clockwise rotation
            layer_elements = (
                layer_elements[rotation_steps:] +
                layer_elements[:rotation_steps]
            )

            index = 0

            # Fill back

            # Top row
            for col in range(left_col, right_col):
                grid[top_row][col] = layer_elements[index]
                index += 1

            # Right column
            for row in range(top_row, bottom_row):
                grid[row][right_col] = layer_elements[index]
                index += 1

            # Bottom row
            for col in range(right_col, left_col, -1):
                grid[bottom_row][col] = layer_elements[index]
                index += 1

            # Left column
            for row in range(bottom_row, top_row, -1):
                grid[row][left_col] = layer_elements[index]
                index += 1

        return grid
